server/protocols/webadmin-integration-test/webadmin-integration-test-common/src/main/java/org/apache/james/webadmin/integration/ForwardIntegrationContract.java

```diff
@@ -49,53 +48,48 @@
 import org.apache.james.probe.DataProbe;
 import org.apache.james.util.Port;
 import org.apache.james.utils.DataProbeImpl;
-import org.apache.james.utils.SMTPMessageSender;
 import org.apache.james.utils.WebAdminGuiceProbe;
 import org.apache.james.webadmin.WebAdminUtils;
 import org.awaitility.Duration;
-import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Tag;
 import org.junit.jupiter.api.Test;
 
 import io.restassured.RestAssured;
 import io.restassured.specification.RequestSpecification;
 
-public abstract class ForwardIntegrationTest {
+public interface ForwardIntegrationContract {
 
-    private SMTPMessageSender messageSender;
-    private RequestSpecification webAdminApi;
-    private Port jmapPort;
+    class TestSystem {
+        private final Port jmapPort;
+        private final RequestSpecification webAdminApi;
+
+        public TestSystem(GuiceJamesServer jamesServer) {
+            this.webAdminApi = WebAdminUtils.spec(
+                jamesServer.getProbe(WebAdminGuiceProbe.class).getWebAdminPort());
+            this.jmapPort = jamesServer.getProbe(JmapGuiceProbe.class).getJmapPort();
+        }
+    }
 
     @BeforeEach
-    void setUp(GuiceJamesServer jmapServer) throws Exception {
-        messageSender = new SMTPMessageSender(DOMAIN);
+    default void setUp(GuiceJamesServer jmapServer, TestSystem testSystem) throws Exception {
```

Comment:
We usually don't rely on the implementation to call a setUp or do @BeforeEach in the interface but each test method call it.
hierarchy of setUp ethods is a pain because you never know what is called first

Reply:
the executing order is explained clearly in the [documentation](https://junit.org/junit5/docs/current/user-guide/#extensions-execution-order-wrapping-behavior). 

I perfer this way instead of calling `setUp()` in each test because I see it put more thing to write, more verbose which is one of the target Junit5 want to address

Reply:
Here is a part of this documentation:

> @BeforeEach methods are inherited from superclasses as long as they are not overridden. Furthermore, @BeforeEach methods from superclasses will be executed before @BeforeEach methods in subclasses.

That means the class has to know there's a default setUp method in the interface to prevent overriding it by accident.

It also means you can't add a default setUp method in the interface without checking for every existing implementations (because it would automatically override it).

Also, if you have a @BeforeEach method in the class that creates the system under test, the @BeforeEach method in the interface can't use it to create a context (like create some users) because it's not initialized yet.

Either I don't understand the documentation or it's just something we don't want to use because of all the shortcomings and pitfalls.

WDYT?

Reply:
Okay, I'll extract the setUp step to another test extension. That would make things don't mess with beforeEach ordering. It's nicer!